src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/MsgPackTraceEventSession.cs

```diff
@@ -0,0 +1,166 @@
+                return eventData;
+            }
+
+            byte[] managedArray = new byte[data.EventDataLength];
+            Marshal.Copy(data.DataStart, managedArray, 0, data.EventDataLength);
+
+            try
+            {
+                var obj = MessagePackSerializer.UnpackMessagePackObject(managedArray);
+                if (obj.IsDictionary)
+                {
+                    foreach (var kv in obj.AsDictionary())
+                    {
+                        var key = kv.Key.AsString();
+                        object val = null;
+                        switch (Type.GetTypeCode(kv.Value.UnderlyingType))
+                        {
+                            case TypeCode.String:
+                                val = kv.Value.AsString();
+                                break;
+
+                            case TypeCode.Boolean:
+                                val = kv.Value.AsBoolean();
+                                break;
+
+                            case TypeCode.Decimal:
+                                val = kv.Value.AsInt64();
+                                break;
+
+                            case TypeCode.Int16:
+                                val = kv.Value.AsInt16();
+                                break;
+
+                            case TypeCode.Int32:
+                                val = kv.Value.AsInt32();
+                                break;
+
+                            case TypeCode.Int64:
+                                val = kv.Value.AsInt64();
+                                break;
+
+                            case TypeCode.UInt16:
+                                val = kv.Value.AsUInt16();
+                                break;
+
+                            case TypeCode.UInt32:
+                                val = kv.Value.AsUInt32();
+                                break;
+
+                            case TypeCode.UInt64:
+                                val = kv.Value.AsUInt64();
+                                break;
+
+                            case TypeCode.Double:
+                                val = kv.Value.AsDouble();
+                                break;
+
+                            case TypeCode.Single:
+                                val = kv.Value.AsSingle();
+                                break;
+
+                            default:
+                                reporter.ReportWarning("Unknown type: " + kv.GetType());
+                                val = kv.Value.ToString();
+                                break;
+                        }
+                        eventData.Payload[key] = val;
+                    }
+                }
+                else
+                {
+                    eventData.Payload["Message"] = obj.ToString();
+                }
+
+            }
+            catch (ArgumentNullException ex)
+            {
+                reporter.ReportProblem("MsgPack decoding failed: " + ex.ToString());
```

Comment:
See comment above about reporting error vs warning if a single event fails to decode